src/azul/plugins/repository/tdr.py

```diff
@@ -113,30 +115,40 @@ def _user_authenticated_tdr(self,
                                   type(authentication))
         return tdr
 
-    def list_sources(self,
-                     authentication: Optional[Authentication]
-                     ) -> list[TDRSourceRef]:
+    _T = TypeVar('_T')
+
+    def _auth_fallback(self,
+                       authentication: Optional[Authentication],
+                       tdr_callback: Callable[[TDRClient], _T]
+                       ) -> _T:
         tdr = self._user_authenticated_tdr(authentication)
-        configured_specs_by_name = {spec.name: spec for spec in self.sources}
-        # Filter by prefix of snapshot names in an attempt to speed up the
-        # listing by limiting the number of irrelevant snapshots returned. Note
-        # that TDR does a substring match, not a prefix match, but determining
-        # the longest common substring is complicated and, as of yet, I haven't
-        # found a trustworthy, reusable implementation.
-        filter = longest_common_prefix(configured_specs_by_name.keys())
         try:
-            snapshots = tdr.snapshot_names_by_id(filter=filter)
+            result = tdr_callback(tdr)
         except UnauthorizedError:
             if tdr.is_registered():
                 raise
             else:
                 # Fall back to anonymous access if the user has authenticated
                 # using an unregistered account. The call to `reject` protects
-                # against infinite recursion in the event that the public
+                # against infinite looping in the event that the public
                 # service account erroneously isn't registered.
                 reject(authentication is None)
-                return self.list_sources(None)
+                return self._auth_fallback(None, tdr_callback)
```

Comment:
If find this recursion rather difficult to reason about. How about the version below? It's simpler because it unrolls the recursion and it needs no `reject`. The unrolling incurs some duplication but I find that OK since the "body" is now encapsulated in the callback.

```diff
Index: src/azul/plugins/repository/tdr.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/repository/tdr.py b/src/azul/plugins/repository/tdr.py
--- a/src/azul/plugins/repository/tdr.py	(revision 195fa8ec235e85fdabb9f58a4ce87eb12d7352c7)
+++ b/src/azul/plugins/repository/tdr.py	(date 1684301861379)
@@ -27,7 +27,6 @@
     CatalogName,
     cache_per_thread,
     config,
-    reject,
     require,
 )
 from azul.auth import (
@@ -88,6 +87,9 @@
         return str(drs_uri.path).strip('/')
 
 
+T = TypeVar('T')
+
+
 @attr.s(kw_only=True, auto_attribs=True, frozen=True)
 class TDRPlugin(RepositoryPlugin[SOURCE_SPEC, SOURCE_REF, BUNDLE_FQID]):
     _sources: Set[TDRSourceSpec]
@@ -115,27 +117,21 @@
                                   type(authentication))
         return tdr
 
-    _T = TypeVar('_T')
-
     def _auth_fallback(self,
                        authentication: Optional[Authentication],
-                       tdr_callback: Callable[[TDRClient], _T]
-                       ) -> _T:
+                       tdr_callback: Callable[[TDRClient], T]
+                       ) -> T:
         tdr = self._user_authenticated_tdr(authentication)
         try:
-            result = tdr_callback(tdr)
+            return tdr_callback(tdr)
         except UnauthorizedError:
-            if tdr.is_registered():
+            if authentication is None or tdr.is_registered():
                 raise
             else:
-                # Fall back to anonymous access if the user has authenticated
-                # using an unregistered account. The call to `reject` protects
-                # against infinite looping in the event that the public
-                # service account erroneously isn't registered.
-                reject(authentication is None)
-                return self._auth_fallback(None, tdr_callback)
-        else:
-            return result
+                # Fall back to anonymous access if the request is authenticated
+                # using an unregistered account.
+                tdr = self._user_authenticated_tdr(None)
+                return tdr_callback(tdr)
 
     def list_sources(self,
                      authentication: Optional[Authentication]
```

Type variables shouldn't be imported so there is no point in making them protected or a class attribute.